Update font.py

LaureFR38 opened the pull request:
Today, when we created a new font, we must enter the kerning values ​​in a JSON file. These values ​​are approximate and are in mm.

From a TTF file it is possible to generate an SVG Font (with FontForge for example).

Extract from a file:
\<glyph glyph-name = "T" unicode = "T" horiz-adv-x = "60" d = "..............." />
\<glyph glyph-name = "a" unicode = "a" horiz-adv-x = "40" d = "..............." />
\<hkern g1 = "T" g2 = "s" k = "7" />

So, this file contains:
- horiz-adv-x: This value is different from that calculated automatically by Ink / Stitch
- k: The amount to decrease the spacing between the two glyphs in the kerning pair.

The unit of these values ​​is the pixel.

So, it is possible to create a font with perfect kerning by modifying the JSON as follows:

Extract from new font.json:
"letter_spacing": 0,
"kerning_pairs_px": { "Ta":7 },
"horiz_adv_x": { "T": 60 , "a":  40 }

Here is my proposal to read the new JSON file (the old version will still work)

lexelby commented on lib/lettering/font.py:
```diff
@@ -113,6 +113,14 @@ def _check_variants(self):
+    kerning_pairs_px = font_metadata('kerning_pairs_px', {}) #Exemple font.json : "kerning_pairs_px": {"AB":3},
+    horiz_adv_x = font_metadata('horiz_adv_x', {}) # Exemple font.json : "horiz_adv_x": {"A":49},
+    horiz_adv_x_default = font_metadata('horiz_adv_x_default', 0) # Exemple font.json : "horiz_adv_x_default" : 45,
+    if version == "2":
```
This won't work the way you're hoping it will, unfortunately. :(  `version` is a `property` object, so it will definitely never `== "2"`.

Instead, you'll need to check the `version` where you're using `word_spacing`.

LaureFR38 replied:
Definitely I'm angry with the version variable. It is again a mistake of inattention.

lexelby commented on lib/lettering/font.py:
```diff
@@ -206,13 +214,20 @@ def _render_glyph(self, glyph, position, character, last_character):
-            position.x += self.letter_spacing + self.kerning_pairs.get(last_character + character, 0) * PIXELS_PER_MM
-
+            if self.version == "2":
+                position.x += glyph._min_x - self.kerning_pairs_px.get(last_character + character, 0)  
```
The prefix `_` in `_min_x` is a way of saying "this is designed only to be used by methods in the Glyph class", a.k.a. a private member variable.  I didn't think we'd need `min_x` outside the Glyph class when I wrote it.  Now it's clear that we need it, so I think we should change the name of the member variable to `min_x`, without the underscore prefix.  Same for `max_x` and the others.

LaureFR38 replied:
Ok. I have to modify or is it you?

lexelby replied:
Go right ahead.  No need to ask me first before changing things.  :)

lexelby replied:
I think it would make sense for it to be part of this pull request.  I might have a chance to do it this evening, or you can -- whoever gets to it first :)

lexelby replied:
Done!

lexelby replied:
Here is something I don't understand: why has this changed to subtracting the `self.kerning_pairs_px` value?  Is this another thing I don't understand about fonts?

LaureFR38 replied:
@lexelby 
The values ​​entered in <hkern g1 = "T" g2 = "s" k = "7" /> are values ​​to be subtracted. I didn't invent the rule: https://www.w3.org/TR/SVG11/fonts.html#HKernElement
If the spacing between letters needs to be made bigger there will be I imagine <hkern g1 = "T" g2 = "s" k = "-7" />. But I think it doesn't happen often.
Did I understand your question correctly?

LaureFR38 replied:
And concerning min_x: I add it before moving the letter because it is to take into account the margin in the drawing of the letter. With respect to point 0 the letter can start at 5 or -5. The letters have a defined place in the drawing that's important.
Then to calculate the position of x for the next letter I have to remove the min_x margin because the horizontal adv is calculated from point 0 of the drawing. You have to look at the fontforge software to understand what is going. 
You can find some frontforge print screens here which explains a bit : =on.https://github.com/inkstitch/inkstitch/issues/809

lexelby replied:
Okay!  Then let's have our kerning work this way too, both for `kerning_pairs` and `kerning_pairs_px`.

I wonder if we should just make `kerning_pairs` be in pixels.....

lexelby reviewed (commented):
So far so good! :+1: 

I know we discussed the `version` thing in DM, and I thought it was a good idea at the time.  But now that I think about it, I wonder if maybe we should not add a `version`, to make the code simpler.  The Lettering extension is still very new, and I think it may be okay if we don't make your changes backward-compatible.  The changes you've made here make Lettering closer to a "real" font system like TrueType, which is a very good thing in my opinion.

Instead, we can just fix the `font.json` for the fonts we already have.  I think that may mean converting `kerning_pairs` to be in pixels instead of millimeters and make them negative(?).  I don't think anyone is distributing Ink/Stitch fonts outside of Ink/Stitch as third-party add-ons, so that won't be an issue.

What do you think?

LaureFR38 commented:
sqd

LaureFR38 commented:
For the fonts we already have, only the DejaVu font has a full json file. And with @Augusa we are already in the process of updating the json file and the ->.svg file. 
The only problem will be for people who have created new fonts. Perhaps during the next version it will be necessary to specify that a new version of the json has been made and that it is not backward-compatible.

Confirm me that I can delete "version".

kaalleen commented:
I think it is possible that people created their own fonts without us knowing about it and it should be backwards compatible.
My proposal would be to check e.g. if 'horiz_adv_x' is present in the file and if so, we know it is "version 2".
What do you think?

LaureFR38 commented:
When Lex say "make the code simple", I don't know if it's about the json file or the python...

lexelby commented:
That makes sense, @kaalleen, but I'm just worried about having a bunch of confusing `if` statements...

Lettering has always been a very very alpha-level feature.  Since it's still so new, I would rather we get it right and potentially inconvenience some folks that may have secretly developed fonts without our knowledge.   I honestly don't think there are many, if any at all.

Whatever way we choose, like @LaureFR38 says, let's just include a good description of what has changed so that people know how to fix their fonts.

> Confirm me that I can delete "version".

Yup, go right ahead.  The nice thing about doing this here on GitHub is that there is always an "undo" button if needed :)

lexelby commented:
> When Lex say "make the code simple", I don't know if it's about the json file or the python...

I was speaking about the Python, but now that you mention it, the JSON should be simple too if possible :)

lexelby commented:
I rebased and force-pushed to bring in the build fixes from #928.

kaalleen commented:
Great to see this is working now. Sorry, I really wasn't aware what the problem was.

We should not forget to update the font files themselves. Would be great for testing too.

LaureFR38 commented:
The reverse direction does not work for fonts that do not specifically have a right-left (although it was originally just a copy of the left-right). To correct this it is necessary to make some modifications in the files font_variant.py in font.py (look for #TODO). Unfortunately I don't have enough skills to do it myself and I don't have an environment to test my code.

LaureFR38 commented:
I have finished adding the 3 new fonts from the French facebook group: Baumans, Chopin, Lobster. 
I updated the kerning of DejaVu, TTdirector, TTmaster. 
GenevaSimple, GenevaRounded, SmallFont and MediumFont don't have kerning but the files had to be updated (svg + json) to work with the new version.

Before each font had 4 direction files. Often times it was a simple copy of the left-right file. 
When I had to update the fonts I had to sort to find out which files were copies or not so that I wouldn't have to modify each file unnecessarily. So I deleted the copy files.
Now we have to make a little change in the code. If the variant is not found use the left right file by default.
I put a #TODO in font.py and font_variant.py and lettering.py where I think it needs to be changed. (Unfortunately I don't know how to do it and I don't have an environment to test the code)
Until then the reverse direction does not work for fonts that do not have a specific right-left (although it was originally just a copy of the left-right)

LaureFR38 commented:
For information :
Baumans : left-right right-left
Chopin : left-right
DejaVu : left-right right-left up-down down-up
GenevaSimple : left-right
GenevaRounded : left-right
Lobster ; left-right
MediumFont : left-right right-left

kaalleen commented:
Sorry for the late comment on my update. I somehow had troubles to login due to my mail provider.

I had a hard time trying to mannually install this branch, because I switched evreything to python3 already. Somehow I always run into troubles while installing, so I finally gave up. Which means I cannot test anything from this branch without uploading it to here.

Hm, I think you are right. It might be worth to check if any variant exist before we exit with an error.

LaureFR38 commented:
@kaalleen I don't know if Lex told you but I would have loved to have an official version for the 25th, a Christmas present for our users with the new lettering. But I think it's too short for you to merge and release the new version. Else for the new year, maybe? What do you think ?

kaalleen commented:
Oh, I don't know. Sounds tempting ;)

Until now, I haven't looked into this branch much. I didn't know what exactly we were trying to achieve ... well - and the time factor - and the python3 issue on my side.

Having the font files definitely will help me to dive into it. Let's try ;)

LaureFR38 commented:
@kaalleen 
I spent a lot of time checking that the new package worked well with all fonts (new and old). I also ask @Augusa
(creator of DejaVu) to test.

The novelty is that the letters are spaced correctly as the creator of the originel font had wanted. It's quite subtle ... But it's essential for handwritten fonts like Chopin and Lobster.

I generate the json with a php script (can't do that manually) with the svg file that fontforge gives me.
In the future, Inkstitch should directly read this svg file. There would therefore be one more file in the police directory. But I don't have the skills to do the python code that does this. So for now if you create a new font you have to tell me and I'll give you the json that goes with it (if you want a nice letter spacing because it still works without).

kaalleen commented:
@LaureFR38 In lettering.py you write you want to uncheck back_and_forth if it is not optimized for this direction?

Is it still valid if we have a fallback for the font variants?

kaalleen commented:
Are there any other issue which we have to be solved with the font spacing, etc before the release?

I think the font creator tool will have to wait until an other release ;) But it's interesting to read your information and shows me where we are going. Maybe we could create a side repository in the inkstitch organisation here https://github.com/inkstitch where you can upload your php script. It might be useful.

LaureFR38 commented:
@kaalleen 
I think.
If the creator has specifically decided to disable for his font, I am thinking in particular of the 2 handwritten fonts made by our French friends then you have to disable. Otherwise for the fonts which have only one file (so when one cannot find variant and reverse variant) it is necessary to uncheck the box because the font has not been optimized for both directions, but the user can check if it really wants to do both ways.

kaalleen commented:
Oh sorry. I haven't read this correctly. It not to disable the checkbox but just to uncheck. I see now.

LaureFR38 commented:
> It not to disable the checkbox but just to uncheck. I see now.
@kaalleen 
That's it but I need both depending on the case

LaureFR38 commented:
@kaalleen 

> Are there any other issue which we have to be solved with the font spacing, etc before the release?

There was a TODO in font.py but I think it is unnecessary now with your modification.

> Maybe we could create a side repository in the inkstitch organisation here https://github.com/inkstitch where you can upload your php script. It might be useful.

I am ashamed to show my php script it is very poorly coded. We will discuss it again.

kaalleen commented:
> I am ashamed to show my php script it is very poorly coded. We will discuss it again.

Haha, ok. I know this feeling ;) But @lexelby encouraged me and told me it's not all to bad what I am doing (although sometimes it is ;))

I am seeing it lettering.py that you already check if the font creator has intended the back and forth option. Are you planing to remove this entirely and only check for the file?

I haven't come up with an useful idea yet. Just reading through the code.

LaureFR38 commented:
> I am seeing it lettering.py that you already check if the font creator has intended the back and forth option. Are you planing to remove this entirely and only check for the file?


I would like :

****
if the font does not have a reverse variant:
	self.back_and_forth_checkbox.uncheck

if font.reversible:
	self.back_and_forth_checkbox.Enable()
else:
	self.back_and_forth_checkbox.Disable()
****
    

>I haven't come up with an useful idea yet. Just reading through the code.

I don't understand.

kaalleen commented:
The build isn't working on Linux so far and we will be in need of more testing also on macOS and Windows.
I guess I am sorry, I don't think we can get this out there in just one day.

LaureFR38 commented:
> The build isn't working on Linux so far and we will be in need of more testing also on macOS and Windows.
> I guess I am sorry, I don't think we can get this out there in just one day.


OK no problem.

kaalleen commented:
```
if the font does not have a reverse variant:
self.back_and_forth_checkbox.uncheck

if font.reversible:
self.back_and_forth_checkbox.Enable()
else:
self.back_and_forth_checkbox.Disable()
```

It looks like the option reversible is set by the author. We can use it as well to decide if the checkbox should be disabled. Would this be sufficient or do we need to check for the files.

LaureFR38 commented:
@kaalleen 

> It looks like the option reversible is set by the author. We can use it as well to decide if the checkbox should be disabled. Would this be sufficient or do we need to check for the files.

I sent the way I thought I would do things. Tell me if it's okay with you and if you understand why I'm asking this.

LaureFR38 commented:
@kaalleen 

I think I was wrong when I requested a TODO in font_variant.py regarding opening the svg file (or its variant). I think it's in font.py and it's a bit more complicated than I thought.

LaureFR38 commented:
@kaalleen 
Can you write me an email at aliaslaure@gmail.com

lexelby commented:
Oh yes, just as @kaalleen says, I want Ink/Stitch to be a safe project for folks of all skill levels to learn on.  No need to feel ashamed about anything :)

@kaalleen I'm not sure about your recent changes on the font variant code.  The way it was before, Font.get_variant handled falling back to left-to-right if a variant file did not exist.  Was that not working properly?

lexelby commented:
Also, it should be okay if a font doesn't have the left-to-right variant if the JSON has default_variant set to another variant.

lexelby commented:
Regarding the PHP script, actually, could we put it in this repository, maybe in the bin/ directory?  @LaureFR38 would it be okay if we translate it to Python?  My PHP skills are a bit rusty but I could probably do it.  I just want to avoid introducing yet another language to the project :)

LaureFR38 commented:
> Regarding the PHP script, actually, could we put it in this repository, maybe in the bin/ directory? @LaureFR38 would it be okay if we translate it to Python? My PHP skills are a bit rusty but I could probably do it. I just want to avoid introducing yet another language to the project :)

The php script browses the svg file from Fontforge and populates the json. If it's to redo the php script in python then in this case you might as well put the script in font.py and no longer need the json about kerning. I don't know in terms of performance if it is faster to browse a json or an svg to find the pairs of characters for the kerning. This should not slow down the generation of the text because it is already long enough ...

LaureFR38 commented:
> @kaalleen I'm not sure about your recent changes on the font variant code. The way it was before, Font.get_variant handled falling back to left-to-right if a variant file did not exist. Was that not working properly?


If it works and that's the problem! I wanted to change the working principle to be able to delete files which are unnecessary copies. And when I delete the right-left file, the right-left return no longer works. We said that we left the choice to the user to want the right left return even if the file does not exist. So I wanted to make him believe that the file existed. That's why Keellin made these changes. But it may not be the best method.

kaalleen commented:
> @kaalleen I'm not sure about your recent changes on the font variant code. The way it was before, Font.get_variant handled falling back to left-to-right if a variant file did not exist. Was that not working properly?

I think you are right. I made the changes without closely looking into the other files. After doing so I expected you to say something like that ;)

I am a bit puzzled about what exactly we are aiming for.

Let's try to summarize:

* Font author doesn't want the font to be reversible
    * Checkbox disabled
    * Checkbox off
    * Default font variant file is always used
* Font author wants the font to be reversible (doesn't specify)
    * Variant file exists
        * Checkbox enabled
        * Checkbox on (depending on previous user settings)
        * Font variant file for chosen direction is used
    * Variant file does not exist
        * Checkbox enabled
        * Checkbox off (depending on previous user settings)
        * Fallback on any existent font variant file (preferably the default font variant file)

Is that what we want to achieve?

LaureFR38 commented:
Yes if when you say "Fallback on any existent font variant file" you mean that when the checkbox is checked, we use the left-right file to do a right-left on the second line, right ?

I'm sorry I have trouble understanding and expressing myself in English.

kaalleen commented:
Yes, that's what I meant. I am having problems to express myself in English too. Especially when I want to say it fast ;)

Ok, I think I got the idea now.

kaalleen commented:
Ok, I think we are on a good way now - at least it should work the way we wanted it to.
@lexelby could you have a look at the code?

kaalleen commented:
@lexelby I am not sure, but we might still have issues with the build process.
When I tried the downloaded version I had an issue which was described by other linux users before (with any Ink/Stitch function).
Ink/Stitch closed immediately after calling the extensions. I have checked for the wayland issue, have checked for permissions and have had a look into the system logs. But I forgot the error from the system logs because I got distracted with other things which I had to do :D I will have to download the updated build again to check again.

Could you try the build on macOS and Linux and see if you encounter the similar issues?

LaureFR38 commented:
> Could you try the build on macOS and Linux and see if you encounter the similar issues?

@Marif91 is on macOS Catalina. She told me that she is ok if we need to install something and try ... 
Tell me what to do.

kaalleen commented:
Oh, that's great. I would like her to download the macOS version from here: https://github.com/inkstitch/inkstitch/releases/tag/dev-build-kerning and see how it is working.

Windows needs testing too. If you are using Windows it'd be nice if you could testrun the windows version from the same link.

kaalleen commented:
@lexelby I had a look into the linux build. It's libgeos_c which is causing the problem. Deleting the libgeos files (while having libgeos installed on the computer) solves the problem ... just not the build. I don't know how to properly fix it. We also should have a look into the big sur issue (from previous build), which was missing lib_c as far as I remember.

LaureFR38 commented:
It's ok on Windows and macOs Catalina (thanks @Marif91)
I will have an answer soon regarding macOs Big Sur

lexelby commented on lib/extensions/lettering.py:
```diff
@@ -174,23 +175,19 @@ def on_font_changed(self, event=None):
-                #by default
-            #    self.back_and_forth_checkbox.SetValue(bool(self.settings.back_and_forth))
+            # TODO replace u"←" with something more clever
+            if u"←" in available_variants:
```
Can we add a function like `Font.has_variant()`?  I think that might be cleaner.

kaalleen replied:
Yes, right :) Naming things again. I wasn't happy with that name but didn't know how to call it. Thanks :)

lexelby commented on lib/extensions/lettering.py:
```diff
@@ -173,6 +175,20 @@ def on_font_changed(self, event=None):
+        available_variants = font.available_variants()
+        if font.reversible:
+            self.back_and_forth_checkbox.Enable()
+            if FontVariant.reversed_variant(font.default_variant) in available_variants:
```
I like how you used `reversed_variant` here, but I think we might need something even more generalized.  Eventually the Lettering extension will allow vertical text as well, and then we'll need to set the checkbox based on whether the top-to-bottom and bottom-to-top variants are both in the font.

This is starting to get really complicated.  Maybe we should just set this checkbox to unchecked by default and leave it at that?

kaalleen replied:
Maybe it's the default which we shouldn't be using? If we go up and down we still would want to know about the reversed version?

But I see why you think this is going to be complicated. And all this checking and unchecking could become confusing to users who do not know the reason why it's doing it. @LaureFR38 What would you think if we just left it unchecked all the time?

LaureFR38 replied:
I think that if it stays unchecked all the time people won't necessarily think of checking it. Also he will not understand why by checking it in some cases the reverse works well (because the file exists) and in other cases not. That's why I thought that checked or unchecked depending on the font let the user understand that the font was not always intended for the opposite direction.

lexelby commented on lib/lettering/font.py:
```diff
@@ -44,7 +45,11 @@ def getter(self):
             # This may be a font packaged with Ink/Stitch, in which case the
             # text will have been sent to CrowdIn for community translation.
             # Try to fetch the translated version.
-            return _(self.metadata.get(name))
+            original_metadata = self.metadata.get(name)
+            localized_metadata = ""
+            if original_metadata != "":
+                localized_metadata = _(self.metadata.get(name))
+            return localized_metadata
```
@kaalleen I don't think this change does what you might have been thinking, based on the description on your commit.  The crowdin files are actually populated from the `bin/inkstitch-fonts-gettext` script.  I think (and hope) that babel ignores the `_(self.metadata.get(name))` call because it's not a simple string.

I included font names and descriptions in the translation strings in crowdin on purpose so they would get translated.  Do you think we should change that?

kaalleen replied:
No, I think it's a good idea to use crowdin for that. It just returns a lot of stuff if we send an empty string - I don't know for what reason. It works the way I indended it, I have been trying. I don't think using the old or new inkscape version would make a huge difference here. But we wouldn't need to call self.metadata again, just use original_metadata.

lexelby replied:
Oh, I see I didn't read this correctly, sorry about that!  I understand how it works now.  Looks good to me.

lexelby commented on lib/lettering/font.py:
```diff
@@ -103,28 +108,58 @@ def _check_variants(self):
     description = localized_font_metadata('description', '')
-    default_variant = font_metadata('default_variant', FontVariant.LEFT_TO_RIGHT)
-    default_glyph = font_metadata('defalt_glyph', u"�")
+    default_glyph = font_metadata('default_glyph', "�")
```
I think this really does need to be a unicode string?

kaalleen replied:
Oh yes, right. I am using the new version to test, we don't need to do that there.

lexelby commented on lib/lettering/font.py:
```diff
@@ -103,28 +108,58 @@ def _check_variants(self):
 
     name = localized_font_metadata('name', '')
     description = localized_font_metadata('description', '')
-    default_variant = font_metadata('default_variant', FontVariant.LEFT_TO_RIGHT)
```
What are your thoughts behind getting rid of default_variant from the font metadata?  I was thinking we could just let the author decide...  is that not working well?

kaalleen replied:
We are still reading it just a bit later and check if it exists as a file.

lexelby replied:
I see it now!

lexelby commented:
I finally had a chance to look over the code, but I haven't had a chance to check out the geos issue yet.  I'm falling asleep at the keyboard, so I think maybe it might be time to call it a night. ;)  I hope to test again soon.

kaalleen commented:
Wow, thank you for your review. I'll have a look at it after work.

kaalleen commented:
Ok, what about telling the user which font variants are actually available for each specific font?

I updated the code for this idea. It still can be modified.

LaureFR38 commented:
> Ok, what about telling the user which font variants are actually available for each specific font?
> 
> I updated the code for this idea. It still can be modified.


Why not, but I'm not sure the user understands what the arrows mean, doesn't know we're dealing with this in 4 files.
There is already a description parameter for the fonts, we can talk about it there.

LaureFR38 commented:
After discussion with @Kaalleen, here is my proposal:

![image](https://user-images.githubusercontent.com/63351363/103242111-76da5500-4955-11eb-8c30-9d8229c4472d.png)

Augusa commented:
On my point of view it looks really too complicated. Some letters from the font (for exemple ABa) and the name of the font is enough. I hope once INkstitch will have 60 fonts as Stitch Era and it will be impossible to see the whole choice. Better gives a short explanation somewhere how to use the lettering. Users of Inkstitch discover a lot of things and are not so surprised.
I don't know if there is place here for my opinion. Sorry if it is not
________________________________
De : LaureFR38 <notifications@github.com>
Envoyé : lundi 28 décembre 2020 21:42
À : inkstitch/inkstitch <inkstitch@noreply.github.com>
Cc : Augusa <lyogau@hotmail.fr>; Mention <mention@noreply.github.com>
Objet : Re: [inkstitch/inkstitch] Update font.py (#848)


After discussion with @kaalleen<https://github.com/kaalleen>, here is my proposal:

[image]<https://user-images.githubusercontent.com/63351363/103242111-76da5500-4955-11eb-8c30-9d8229c4472d.png>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/inkstitch/inkstitch/pull/848#issuecomment-751858079>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANDK3S5UH7GHO7DJ527YWY3SXDUUXANCNFSM4SKP2P6A>.

LaureFR38 commented:
> Some letters from the font (for exemple ABa) and the name of the font is enough.

I agree that it would be enough to put a few letters in but only if Ink / Stitch were commercial software. Currently there are not many fonts that all have French accented letters for example. I find it good to check for the font we will choose for a project with all the letters. It avoids unpleasant surprises.

LaureFR38 commented:
After chatting with Augusa, I agree with her. The image overloads the display. And the day a font has a lot of character, it will be too big to display.
If you can get an image of the font name in the drop-down list that would be enough to judge the style of the font.

Augusa commented:
I give some more explanations on my view on lettering. I just say some thing on the description. I find it useless in the list. When I want to choose I just need to see the name of the font with the aspect when embroidered. May be It would be nice to have a question mark that leads to an explanation page on the fonts and their detail with the operation of the boxes grayed out or not.
I think it is normal to search for explanations about the boxes here as everywhere in Inkstitch. Hope it help to clear my ideas.

lexelby commented:
I really don't think we should tell the user which variants are available.  As @LaureFR38 says, it's not something the user should need to understand.  Some fonts will work fine in and direction.  Imagine a font with big letters with fill stitch, for example.

I'm fine with not showing the description to the user.  I included it with the idea that some fonts may be for a special purpose, like the small font I made.

LaureFR38 commented:
@lexelby 
For fonts which do not have a right-left file (and which are not in automatic satin), how to warn the user that the font was not intended from right to left (but that he can do it anyway) He will notice it on his own?

lexelby commented:
The font may still be intended for right-to-left text.  If no changes are required to make the font work well right-to-left, then the author won't include a right-to-left SVG.  The font may nevertheless work fine right-to-left.  

We can't make any assumptions just because the SVG file is not there.  We should just depend on the "reversible" setting from font.json.

lexelby commented:
Imagine a Morse Code font made of satin stitch.  It would work just fine forward or backward without changes.

Augusa commented:
> 
> 
> Imagine a Morse Code font made of satin stitch. It would work just fine forward or backward without changes.

I am surprised. If you use a right to left file and embroider it from left to right you get jumps stitches UNDER The following letter. Of course on modern machines it is possible to cut the thread but it needs a lot of time and often it is not really necessary to cut the thread between small letters when the font is correctly build.

lexelby commented:
Don't forget, it still runs the satins through Auto-Route Satin Columns.

kaalleen commented:
Where is the harm to inform users about the available directions? I myself like to have a look on it. The auto satin does a great job. On morse code it really wouldn't matter at all which direction was created by the font author. This might not be the case for some other fonts and you would want to inspect them more closely when you used the automated back and forward option.

It'd be better to add a tooltipp to inform about the auto-satin option for reversed files but there is an error in wxpython which makes it impossible to add tooltipps within staticboxes.

lexelby commented:
Trying the user which direction the font is designed to work in is fine with me.  I just want to have the font author explicitly say in the font.json which directions the font is designed to work with, rather than assuming based on which variant files exist.

LaureFR38 commented:
> Trying the user which direction the font is designed to work in is fine with me. I just want to have the font author explicitly say in the font.json which directions the font is designed to work with, rather than assuming based on which variant files exist.

Ok i will update the json descriptions.

kaalleen commented:
ok

lexelby commented:
"trying" up there should have been "telling"

kaalleen commented:
@lexelby do you think this is ready to be merged?

lexelby commented:
Looks great to me!  I tested it out and I love it!  Those new fonts are _incredible_!!  This really brings Ink/Stitch's lettering to the next level.  Thank you all for your work on this!

@LaureFR38 I'll leave the final decision to you: are we ready to merge and release a new version?

LaureFR38 commented:
> @LaureFR38 I'll leave the final decision to you: are we ready to merge and release a new version?

Yes I agree to merge my request. For the new version if we can wait a little longer. @kaalleen is working on the generation of the json and it would be nice to have it in the new version. Besides, my French friends still have new fonts in the works, maybe some will be finished this month... maybe.....

lexelby commented:
Sounds good to me!
